internal/storage/overlay/store.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2021-2023 Zenauth Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+package overlay
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	runtimev1 "github.com/cerbos/cerbos/api/genpb/cerbos/runtime/v1"
+	"go.uber.org/zap"
+
+	"github.com/cerbos/cerbos/internal/compile"
+	"github.com/cerbos/cerbos/internal/config"
+	"github.com/cerbos/cerbos/internal/engine"
+	"github.com/cerbos/cerbos/internal/namer"
+	"github.com/cerbos/cerbos/internal/policy"
+	"github.com/cerbos/cerbos/internal/schema"
+	"github.com/cerbos/cerbos/internal/storage"
+	"github.com/sony/gobreaker"
+)
+
+const (
+	DriverName = "overlay"
+)
+
+var _ Overlay = (*Store)(nil)
+
+func init() {
+	storage.RegisterDriver(DriverName, func(ctx context.Context, confW *config.Wrapper) (storage.Store, error) {
+		conf := new(Conf)
+		if err := confW.GetSection(conf); err != nil {
+			return nil, fmt.Errorf("failed to read overlay configuration: %w", err)
+		}
+
+		return NewStore(ctx, conf, confW)
+	})
+}
+
+func NewStore(ctx context.Context, conf *Conf, confW *config.Wrapper) (*Store, error) {
+	getStore := func(key string) (storage.SourceStore, error) {
+		cons, err := storage.GetDriverConstructor(key)
+		if err != nil {
+			return nil, fmt.Errorf("unknown storage driver [%s]", key)
+		}
+
+		store, err := cons(ctx, confW)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create overlay store: %w", err)
+		}
+
+		sourceStore, ok := store.(storage.SourceStore)
```

Comment:
Why is it necessary for the store to be a `SourceStore`? Isn't it enough to implement the `Store` interface? 

In my mind, I was thinking that the `overlay` driver could implement all the relevant interfaces (`SourceStore`, `BinaryStore`, and `Reloadable`). Then, based on which driver was active, each of those methods could either call the corresponding method on the underlying driver or return an error.

Reply:
I have been thinking about this a bit more and, at the cost of complicating things a little bit, what if the `overlay` driver always exposed the same interfaces as the `base` driver?  When the fallback driver is active -- which should not happen quite so often anyway -- it's OK to return errors for methods that couldn't be invoked or emulated using it.

Reply:
My thinking was that we weren't likely to be configuring overlays with `BinaryStore`s (assuming the only implementations of them were bundle/cloud related). Asserting on `SourceStore` shows intent and simplifies the interface methods by negating the need for type assertions/switches in each. Is this likely to be used with `BinaryStore` drivers?

Omitting `Reloadable` was an arbitrary decision (similar to the decision to avoid any mutable operations) but I realise now that this is an oversight given all(most?) stores seem to support it!

> what if the overlay driver always exposed the same interfaces as the base driver?

This would make sense to me in the scenario where driver interfaces frequently varied. I think I'm struggling to see the pay-off now given I'm stuck in a "`SourceStore` only" mindset 🤔.

Reply:
I think `BinaryStore` is a prime candidate for having a fallback driver configured in case of connectivity issues to the cloud. I expect that some users would want to configure `bundle` as the primary store but have `git` or `disk` as fallback drivers for redundancy.

Reply:
In that case, enforcing the base driver interface makes sense 👍